scheduler/src/cook/compute_cluster.clj

```diff
@@ -156,3 +160,339 @@
+                          db))]
+    (map-from-vals :compute-cluster-config/name configs)))
+
+(defn db-config-ents->configs
+  "Convert cluster configuration database entities to plain configurations"
+  [current-db-config-ents]
+  (map-vals compute-cluster-config-ent->compute-cluster-config current-db-config-ents))
+
+(defn get-db-configs
+  "Get the current dynamic cluster configurations from the database"
+  [db]
+  (->> (get-db-config-ents db)
+       (map-vals compute-cluster-config-ent->compute-cluster-config)))
+
+(defn get-db-configs-latest
+  "Delegates to get-db-configs with a fresh db"
+  []
+  (get-db-configs (d/db datomic/conn)))
+
+;TODO: when all clusters have state, change the ComputeCluster protocol to add state operations
+(defn compute-cluster->compute-cluster-config
+  "Calculate dynamic cluster configuration from a compute cluster"
+  [{:keys [state-atom state-locked?-atom name]
+    {{:keys [template base-path ca-cert]} :config} :cluster-definition}]
+  {:name name
+   :template template
+   :base-path base-path
+   :ca-cert ca-cert
+   :state @state-atom
+   :state-locked? @state-locked?-atom})
+
+;TODO: in the future all clusters will be dynamic and we shouldn't need this
+(defn get-dynamic-clusters
+  "Get the current in-memory dynamic clusters"
+  []
+  (->> @cluster-name->compute-cluster-atom
+       (filter (fn [[_ cluster]] (:dynamic-cluster-config? cluster)))
+       (into {})))
+
+(defn get-in-mem-configs
+  "Get the current in-memory dynamic cluster configurations"
+  []
+  (->> (get-dynamic-clusters)
+       (map-vals compute-cluster->compute-cluster-config)))
+
+(defn compute-current-configs
+  "Synthesize the current view of cluster configurations by looking at the current configurations in the database
+  and the current configurations in memory. Alert on any inconsistencies. Database wins on inconsistencies."
+  [current-db-configs current-in-mem-configs]
+  (let [[only-db-keys only-in-mem-keys both-keys] (util/diff-map-keys current-db-configs current-in-mem-configs)]
+    (doseq [only-db-key only-db-keys]
+      (when (not= :deleted (-> only-db-key current-db-configs :state))
+        (log/info "Database cluster configuration is missing from in-memory configs."
+                  "Cluster is only in the database and is not deleted."
+                  "This is normal at startup, but shouldn't happen otherwise."
+                  {:cluster-name only-db-key :cluster (current-db-configs only-db-key)})))
+    (doseq [only-in-mem-key only-in-mem-keys]
+      (when (not= :deleted (-> only-in-mem-key current-in-mem-configs :state))
+        (log/error "In-memory cluster configuration is missing from the database."
+                   "Cluster is only in memory and is not deleted."
+                   "This should not have happened! A database update was missed!"
+                   "The cluster configuration will be added to the database on the next update."
```

Comment:
If its from the dynamic-cluster loop, then yes, it'll be repaired the next time the query happens. If it came from a REST API, it won't. So this seems wrong. Maybe:

"If this cluster was added in by the <<fill in function name here for cluster refresh>> this should be fixed by the next time it runs. IF it was added via REST, then they should be retrying"

Reply:
Moot now that the database write goes first.